relations/jf/src/lib.rs

```diff
@@ -26,18 +27,16 @@ pub fn generate_srs<R: CryptoRng + RngCore>(
 }
 
 /// Common API for all relations.
-pub trait Relation: Default {
-    /// Public input to the relation. Must be marshallable.
-    type PublicInput: Marshall;
-    /// Private input to the relation.
-    type PrivateInput;
-
-    /// Constructs new relation object from public and private inputs.
-    fn new(public_input: Self::PublicInput, private_input: Self::PrivateInput) -> Self;
-
-    /// Include this relation in the circuit.
-    fn generate_subcircuit(&self, circuit: &mut PlonkCircuit<CircuitField>) -> PlonkResult<()>;
+pub trait Relation {
+    /// Include this relation in the circuit and return all variables that are needed for outside circuit.
+    fn generate_subcircuit(
+        &self,
+        circuit: &mut PlonkCircuit<CircuitField>,
+    ) -> PlonkResult<Vec<Variable>>;
```

Comment:
I guess that this will be a pain eventually - it's unstructured, var-length sequence of plain `usize`, which might bring some stupid errors; also, it's hard to tell which variables are needed for outside circuit in general;

I think that the ideal thing would be to return a mirror type with all fields of `Variable` type

Reply:
This return type here is needed e.g. in withdraw. I don't see a clear API how to combine subcircuits that are not self contained in clear API.

Reply:
outdated